lib/mongo/uri/options_mapper.rb

```diff
@@ -584,31 +655,38 @@ def revert_w(value)
       # Parses the zlib compression level.
       #
       # @param [ String ] name Name of the URI option being processed.
-      # @param value [ String ] The zlib compression level string.
+      # @param [ String | Integer ] value The zlib compression level string.
       #
       # @return [ Integer | nil ] The compression level value if it is between -1 and 9 (inclusive),
       #   otherwise nil (and a warning will be logged).
       def convert_zlib_compression_level(name, value)
-        if /\A-?\d+\z/ =~ value
-          i = value.to_i
-
-          if i >= -1 && i <= 9
-            return i
-          end
+        i = if value.is_a?(String) && /\A-?\d+\z/ =~ value
+          value.to_i
+        elsif value.is_a?(Integer)
+          value
         end
 
-        log_warn("#{value} is not a valid zlibCompressionLevel")
-        nil
+        if i && (i >= -1 && i <= 9)
```

Comment:
This also I think mishandles input which is neither string nor integer.

Reply:
Again, when it's neither a string nor integer, i will be nil and exit (this time on the first part of the if condition)

Reply:
Again, when it's neither a string nor integer, i will be nil and exit (this time on the first part of the if condition)